tests/components/onvif/test_config_flow.py

```diff
@@ -0,0 +1,508 @@
+        services.append(service)
+    mock_discovery.return_value = services
+
+
+def setup_mock_camera(mock_camera):
+    """Prepare mock HASS camera."""
+    mock_camera.async_initialize.return_value = Future()
+    mock_camera.async_initialize.return_value.set_result(True)
+
+    def mock_constructor(hass, config):
+        """Fake the controller constructor."""
+        return mock_camera
+
+    mock_camera.side_effect = mock_constructor
+
+
+async def setup_onvif_integration(
+    hass, config=None, options=None, unique_id=MAC, entry_id="1", source="user",
+):
+    """Create an ONVIF config entry."""
+    if not config:
+        config = {
+            config_flow.CONF_NAME: NAME,
+            config_flow.CONF_HOST: HOST,
+            config_flow.CONF_PORT: PORT,
+            config_flow.CONF_USERNAME: USERNAME,
+            config_flow.CONF_PASSWORD: PASSWORD,
+            config_flow.CONF_PROFILE: [0],
+        }
+
+    config_entry = MockConfigEntry(
+        domain=config_flow.DOMAIN,
+        source=source,
+        data={**config},
+        connection_class=config_entries.CONN_CLASS_LOCAL_PUSH,
+        options=options or {},
+        entry_id=entry_id,
+        unique_id=unique_id,
+    )
+    config_entry.add_to_hass(hass)
+
+    with patch(
+        "homeassistant.components.onvif.config_flow.get_device"
+    ) as mock_device, patch(
+        "homeassistant.components.onvif.config_flow.wsdiscovery"
+    ) as mock_discovery, patch(
+        "homeassistant.components.onvif.camera.ONVIFHassCamera"
+    ) as mock_camera:
+        setup_mock_onvif_device(mock_device, two_profiles=True)
+        # no discovery
+        mock_discovery.return_value = []
+        setup_mock_camera(mock_camera)
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+    return config_entry
+
+
+async def test_flow_discovered_devices(hass):
+    """Test that config flow works for discovered devices."""
+
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": "user"}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    with patch(
+        "homeassistant.components.onvif.config_flow.get_device"
+    ) as mock_device, patch(
+        "homeassistant.components.onvif.config_flow.wsdiscovery"
+    ) as mock_discovery, patch(
+        "homeassistant.components.onvif.camera.ONVIFHassCamera"
+    ) as mock_camera:
+        setup_mock_onvif_device(mock_device)
+        setup_mock_discovery(mock_discovery)
+        setup_mock_camera(mock_camera)
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], user_input={}
+        )
+
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "device"
+        assert len(result["data_schema"].schema[config_flow.CONF_HOST].container) == 3
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5453535)